src/main/java/org/seqdoop/hadoop_bam/util/SAMFileMerger.java

```diff
@@ -0,0 +1,150 @@
+package org.seqdoop.hadoop_bam.util;
+
+import com.google.common.io.CountingOutputStream;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.cram.build.CramIO;
+import htsjdk.samtools.cram.common.CramVersions;
+import htsjdk.samtools.util.BlockCompressedFilePointerUtil;
+import htsjdk.samtools.util.BlockCompressedStreamConstants;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import org.seqdoop.hadoop_bam.KeyIgnoringAnySAMOutputFormat;
+import org.seqdoop.hadoop_bam.SAMFormat;
+import org.seqdoop.hadoop_bam.SplittingBAMIndex;
+import org.seqdoop.hadoop_bam.SplittingBAMIndexer;
+
+import static org.seqdoop.hadoop_bam.util.NIOFileUtil.asPath;
+import static org.seqdoop.hadoop_bam.util.NIOFileUtil.deleteRecursive;
+import static org.seqdoop.hadoop_bam.util.NIOFileUtil.getFilesMatching;
+import static org.seqdoop.hadoop_bam.util.NIOFileUtil.mergeInto;
+
+/**
+ * Merges headerless BAM or CRAM files produced by {@link KeyIgnoringAnySAMOutputFormat}
+ * into a single file.
+ */
+public class SAMFileMerger {
+
+  private SAMFileMerger() {
+  }
+
+  /**
+   * Merge part file shards produced by {@link KeyIgnoringAnySAMOutputFormat} into a
+   * single file with the given header.
+   * @param partDirectory the directory containing part files
+   * @param outputFile the file to write the merged file to
+   * @param samOutputFormat the format (must be BAM or CRAM; SAM is not supported)
+   * @param header the header for the merged file
+   * @throws IOException
+   */
+  public static void mergeParts(final String partDirectory, final String outputFile,
+      final SAMFormat samOutputFormat, final SAMFileHeader header) throws IOException {
+
+    // First, check for the _SUCCESS file.
+    final String successFile = partDirectory + "/_SUCCESS";
+    final Path successPath = asPath(successFile);
+    if (!Files.exists(successPath)) {
+      throw new NoSuchFileException(successFile, null, "Unable to find _SUCCESS file");
+    }
+    final Path partPath = asPath(partDirectory);
+    final Path outputPath = asPath(outputFile);
+    if (partPath.equals(outputPath)) {
+      throw new IllegalArgumentException("Cannot merge parts into output with same " +
+          "path: " + partPath);
+    }
+
+    List<Path> parts = getFilesMatching(partPath,
+        "glob:**/part-[mr]-[0-9][0-9][0-9][0-9][0-9]*",
```

Comment:
This glob string should be a constant somewhere as its repeated in several places.